pkg/kubelet/dockertools/exec.go

```diff
@@ -122,47 +131,77 @@ func (*NativeExecHandler) ExecInContainer(client DockerInterface, container *doc
 		return fmt.Errorf("failed to exec in container - Exec setup failed - %v", err)
 	}
 
-	// Have to start this before the call to client.StartExec because client.StartExec is a blocking
-	// call :-( Otherwise, resize events don't get processed and the terminal never resizes.
-	kubecontainer.HandleResizing(resize, func(size term.Size) {
-		client.ResizeExecTTY(execObj.ID, int(size.Height), int(size.Width))
-	})
-
-	startOpts := dockertypes.ExecStartCheck{Detach: false, Tty: tty}
-	streamOpts := StreamOptions{
-		InputStream:  stdin,
-		OutputStream: stdout,
-		ErrorStream:  stderr,
-		RawTerminal:  tty,
-	}
-	err = client.StartExec(execObj.ID, startOpts, streamOpts)
-	if err != nil {
-		return err
-	}
+	ch := make(chan error, 1)
+	// Run client.StartExec in a new goroutine, because it is blocking and
+	// we may want to terminate earlier in case of a timeout.
+	go func() {
+		// Have to start this before the call to client.StartExec because client.StartExec is a blocking
+		// call :-( Otherwise, resize events don't get processed and the terminal never resizes.
+		kubecontainer.HandleResizing(resize, func(size term.Size) {
+			client.ResizeExecTTY(execObj.ID, int(size.Height), int(size.Width))
+		})
 
-	ticker := time.NewTicker(2 * time.Second)
-	defer ticker.Stop()
-	count := 0
-	for {
-		inspect, err2 := client.InspectExec(execObj.ID)
-		if err2 != nil {
-			return err2
+		startOpts := dockertypes.ExecStartCheck{Detach: false, Tty: tty}
+		streamOpts := StreamOptions{
+			InputStream:  stdin,
+			OutputStream: stdout,
+			ErrorStream:  stderr,
+			RawTerminal:  tty,
 		}
-		if !inspect.Running {
-			if inspect.ExitCode != 0 {
-				err = &dockerExitError{inspect}
-			}
-			break
+		err = client.StartExec(execObj.ID, startOpts, streamOpts)
+		if err != nil {
+			ch <- err
+			return
 		}
 
-		count++
-		if count == 5 {
-			glog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
-			break
+		ticker := time.NewTicker(2 * time.Second)
+		defer ticker.Stop()
+		count := 0
+		for {
+			inspect, err2 := client.InspectExec(execObj.ID)
+			if err2 != nil {
+				ch <- err2
+				return
+			}
+			if !inspect.Running {
+				if inspect.ExitCode != 0 {
+					err = &dockerExitError{inspect}
+				}
+				break
+			}
+			count++
+			if count == 5 {
+				glog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
+				break
+			}
+			<-ticker.C
 		}
+		ch <- err
+	}()
 
-		<-ticker.C
+	// No timeout, block until exec is finished.
+	if timeout <= 0 {
+		return <-ch
+	}
+	// Otherwise, wait for completion or timeout, whatever happens first.
+	select {
+	case err := <-ch:
+		return err
+	case <-time.After(timeout):
```

Comment:
Is the problem the Docker API which doesn't support cancel or just the Docker client lib? In the later case we should open an issue (or even a PR) to add timeouts.